src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,449 @@
+            // I'm not sure why this signature needs to be a const reference to
+            // make the typechecker happy when the visitor for BaseFactToDLIR
+            // did
+            // not need this.
+            [this, &speaker](const Fact& fact) {
+              auto [fact_predicate, generated_rules] =
+                  FactToDLIR(speaker, fact);
+              DLIRAssertion main_assertion = DLIRAssertion(
+                  PushPrin(std::string("says_"), speaker, fact_predicate));
+              generated_rules.push_back(main_assertion);
+              return generated_rules;
+            },
+            [this,
+             &speaker](const ConditionalAssertion& conditional_assertion) {
+              std::vector<Predicate> dlir_rhs = {};
+              for (auto ast_rhs : conditional_assertion.rhs()) {
+                // extra rule are only generated for facts on the LHS,
+                // so the rules that would be generated from this RHS fact are
+                // not used.
+                auto [dlir_translation, not_used] =
+                    BaseFactToDLIR(speaker, ast_rhs);
+                dlir_rhs.push_back(
+                    PushPrin(std::string("says_"), speaker, dlir_translation));
+              }
+              auto [dlir_lhs, gen_rules] =
+                  FactToDLIR(speaker, conditional_assertion.lhs());
+              auto dlir_lhs_prime =
+                  PushPrin(std::string("says_"), speaker, dlir_lhs);
+              auto dlir_assertion =
+                  DLIRAssertion(DLIRCondAssertion(dlir_lhs_prime, dlir_rhs));
+              gen_rules.push_back(dlir_assertion);
+              return gen_rules;
+            }},
+        assertion.GetValue());
+  }
+
+  std::vector<DLIRAssertion> SaysAssertionToDLIR(
+      const SaysAssertion& says_assertion) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const Assertion& assertion : says_assertion.assertions()) {
+      std::vector<DLIRAssertion> single_translation =
+          SingleSaysAssertionToDLIR(says_assertion.principal(), assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
+    }
+    return ret;
+  }
+
+  std::vector<DLIRAssertion> SaysAssertionsToDLIR(
+      const std::vector<SaysAssertion>& says_assertions) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const SaysAssertion& says_assertion : says_assertions) {
+      auto single_translation = SaysAssertionToDLIR(says_assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
+    }
+    return ret;
+  }
+
+// Queries are like predicates with arity 0. Souffle does not have predicates
+// with arity 0, so we represent them as having one argument which is a
+// constant.
+#define DUMMY_PREDICATE Predicate("grounded_dummy", {"dummy_var"}, kPositive)
```

Comment:
In general, [macros are not encouraged in Google C++]( https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros). Try making this a `static const` global instead, like so:

```C++
static const Predicate kDummyPredicate("grounded_dummy", {"dummy_var"}, kPositive);
```

The compiler will "do the right thing" and inline construction of this global if optimizations are on and it thinks it will help.

Reply:
That doesn't quite compile. I did a similar thing that's probably also fine?